### DISCUSSION TITLE
Multiple activities per project, automatic activity definition

### DISCUSSION THREAD

**Discussion by psionic-k** [alphapapa/activities.el#69] [Ideas]
This is a proposal for project integration, at least what would drive requirements.

Let's say I work on microservices.com and they use a git repository per microservice.  The engineers typically work on 3-4 sets of files depending on which aspect of the service they are working on:

1. Feature work, working on source and unit tests
2. CI & CD, working on Github actions and Nix flakes
3. K8's diagnosis and designing production tests, building and deploying ephemeral debugging containers
4. Network, retry, metrics definitions for assimilation by Linkerd
5. Common data model definition that creates tables and API endpoints through code generation

Since microservices.com has 100 services and each engineer works on at least 10 services, automating the project navigation by automatically generating activities has been requested by the CTO.

The implementation would rely on the current project.  It would be configured with a list of `(NAME PREDICATE ACTIVITY-FUNCTION)`.  If `PREDICATE` passes, `NAME-<PROJECT>` is available and can be generated if it doesn't exist already by calling `ACTIVITY-FUNCTION`

**Comment by alphapapa** [comment]
That's certainly an interesting idea.  What would ACTIVITY-FUNCTION do exactly?

**Comment by psionic-k** [comment]
`ACTIVITY-FUNCTION` would return .. a valid activity or enough data to create one.  If writing those functions is hard, then that's something worth having a macro or library functions for.

**Comment by alphapapa** [comment]
An activity is represented by a window configuration as returned by `window-state-get`, with each window having an associated record that can be used to restore its buffer.  The state can be restored by `window-state-put` (after `activities--bufferize` having recreated its buffers).

A window configuration is a Lisp object, so a function could be written to create one from scratch.  But the intention is for the state to be created by the user's arranging various windows and buffers.  For that to be automated, one could call various `split-window` and `switch-to-buffer` commands, as is shown by the examples in my `yequake` package's documentation.

But if one were sufficiently motivated, one could either write a function to make the window configuration object, or to take a simple, default one and return a copy of it having been modified according to some goal.  I think that would be out of scope for the `activities` library itself, but it could be done by another library and the result could then be used as an argument to appropriate `activities` functions.

**Reply by tomdavey** [reply]
Alphapapa, this is an excellent, generous reply. I learned a lot about the internals of activities.el from it. To that same end, I am reading through this entire repository with an ever-growing appreciation for activities.el.
